scripts/esbuild/internal-platform-client.ts

```diff
@@ -0,0 +1,110 @@
+import type { BuildOptions as ESBuildOptions } from 'esbuild';
+import { replace } from 'esbuild-plugin-replace';
+import fs from 'fs-extra';
+import glob from 'glob';
+import { join } from 'path';
+
+import { getBanner } from '../utils/banner';
+import { BuildOptions, createReplaceData } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { externalAlias, getBaseEsbuildOptions, getEsbuildAliases, getEsbuildExternalModules } from './util';
+
+/**
+ * Create objects containing ESbuild options for the two bundles which need to
+ * be written to `internal/client`. This also performs relevant side-effects,
+ * like clearing out the directory and writing a `package.json` script to disk.
+ *
+ * @param opts build options
+ * @returns an array of ESBuild option objects
+ */
+export async function getInternalClientBundle(opts: BuildOptions): Promise<ESBuildOptions[]> {
+  const inputClientDir = join(opts.srcDir, 'client');
+  const outputInternalClientDir = join(opts.output.internalDir, 'client');
+  const outputInternalClientPolyfillsDir = join(outputInternalClientDir, 'polyfills');
+
+  await fs.emptyDir(outputInternalClientDir);
+  await fs.emptyDir(outputInternalClientPolyfillsDir);
+
+  await copyPolyfills(opts, outputInternalClientPolyfillsDir);
+
+  // write @stencil/core/internal/client/package.json
+  writePkgJson(opts, outputInternalClientDir, {
+    name: '@stencil/core/internal/client',
+    description:
+      'Stencil internal client platform to be imported by the Stencil Compiler and internal runtime. Breaking changes can and will happen at any time.',
+    main: 'index.js',
+    sideEffects: false,
+  });
+
+  const internalClientAliases = getEsbuildAliases();
+  internalClientAliases['@platform'] = join(inputClientDir, 'index.ts');
+
+  const clientExternal = getEsbuildExternalModules(opts, opts.output.internalDir);
+
+  const internalClientBundle: ESBuildOptions = {
+    ...getBaseEsbuildOptions(),
+    entryPoints: [join(inputClientDir, 'index.ts')],
+    format: 'esm',
+    outfile: join(outputInternalClientDir, 'index.js'),
+    platform: 'node',
+    external: clientExternal,
+    alias: internalClientAliases,
+    banner: {
+      js: getBanner(opts, 'Stencil Client Platform'),
+    },
+    plugins: [
+      replace(createReplaceData(opts)),
+      externalAlias('@app-data', '@stencil/core/internal/app-data'),
+      externalAlias('@utils/shadow-css', './shadow-css.js'),
```

Comment:
yeah that's right. I added a separate new bundle here:

https://github.com/ionic-team/stencil/blob/71701bfc35c6c5625337a65ac499fb3843401643/scripts/esbuild/internal.ts#L47-L54

this builds `src/utils/shadow-css.ts` as a separate bundle and writes it to `internal/client/shadow-css.js` so that then the various other modules in `internal/` can all import it.

Rollup's behavior w/r/t this file before was to copy it to each location, which lead to unnecessary duplication. With Esbuild, by contrast, it was getting bundled into all those files (like `internal/client/index.js` and friends) by default.

Reply:
I added the `paths` alias `@utils/shadow-css` to make it a bit easier to intercept this import in a location-independent way (I don't have to worry about where the importing module is relative to `src/utils/shadow-css.ts`)

Reply:
Is this something we'll need to be aware of when comparing bundle sizes/contents during the verification work?

Reply:
@tanner-reits I think so, otherwise a diff would look something like this:
<img width="876" alt="Screenshot 2024-02-12 at 11 57 41 AM" src="https://github.com/ionic-team/stencil/assets/1930213/b1b0566f-ece5-41cd-98f1-e88c3f208a64">
Where `shadow-css` is a file that's only in rollup (right hand side) for the `hydrate` and `testing` internal subpackages